cpp/src/replace/replace.cu

```diff
@@ -87,6 +90,109 @@ __device__ auto get_new_value(cudf::size_type idx,
+  cudf::size_type i = blockIdx.x * blockDim.x + threadIdx.x;
+  while (i < nrows) {
+    bool input_is_valid = true;
+    if (input_has_nulls)
+      input_is_valid = input.is_valid_nocheck(i);
+    if (input_is_valid){
+      int result = get_new_string_value(i, input, values_to_replace, replacement);
+      cudf::string_view output = (result == -1) ? input.element<cudf::string_view>(i) : replacement.element<cudf::string_view>(result);
+      offsets.data<cudf::size_type>()[i] = output.size_bytes();
+      indices.data<cudf::size_type>()[i] = result;
+    } else {
+      offsets.data<cudf::size_type>()[i] = 0;
+      indices.data<cudf::size_type>()[i] = -1;
+    }
+    i += blockDim.x * gridDim.x;
+  }
+}
+
+template<bool input_has_nulls, bool replacement_has_nulls>
+__global__ void replace_strings_second_pass(cudf::column_device_view input,
+                                            cudf::column_device_view replacement,
+                                            cudf::mutable_column_device_view offsets,
+                                            cudf::mutable_column_device_view strings,
+                                            cudf::mutable_column_device_view indices,
+                                            cudf::bitmask_type * output_valid,
+                                            cudf::size_type * __restrict__ output_valid_count) {
+  cudf::size_type nrows = input.size();
+  cudf::size_type i = blockIdx.x * blockDim.x + threadIdx.x;
+  uint32_t active_mask = 0xffffffff;
+  active_mask = __ballot_sync(active_mask, i < nrows);
+  auto const lane_id { threadIdx.x % cudf::experimental::detail::warp_size };
+  uint32_t valid_sum { 0 };
+  while (i < nrows) {
+    bool output_is_valid = true;
+    bool input_is_valid = true;
+    cudf::size_type idx = indices.element<cudf::size_type>(i);
+    uint32_t bitmask = 0xffffffff;
+
+
+    if (input_has_nulls){
+      input_is_valid = input.is_valid_nocheck(i);
+      output_is_valid = input_is_valid;
+    }
+    if (replacement_has_nulls && idx != -1) {
+      output_is_valid = replacement.is_valid_nocheck(idx);
+    }
+    if (output_is_valid) {
+      cudf::string_view output = (idx == -1) ? input.element<cudf::string_view>(i) : replacement.element<cudf::string_view>(idx);
+      std::memcpy(strings.data<char>() + offsets.data<cudf::size_type>()[i],
+                  output.data(),
+                  output.size_bytes());
+    }
+
+    /* output valid counts calculations*/
+    if (input_has_nulls or replacement_has_nulls) {
+
+      bitmask = __ballot_sync(active_mask, output_is_valid);
+
+      if (0 == lane_id) {
+        output_valid[cudf::word_index(i)] = bitmask;
+        valid_sum += __popc(bitmask);
+      }
+    }
+
+
+    i += blockDim.x * gridDim.x;
+    active_mask = __ballot_sync(active_mask, i < nrows);
+  }
+  if (input_has_nulls or replacement_has_nulls) {
+    __syncthreads(); // waiting for the valid counts of each warp to be ready
```

Comment:
Why is a sync threads needed? There is no shared memory in use, and `valid_sum` is a thread local variable. The reduce below will take care of any necessary synchronization.

```suggestion
```